## Let's do this!

Name: William Pereira

First I'll put the description relative to the challenge, them I'll import some libraries and do some programming to answer the questions. 

Linx Impulse is atop the world of e-commerce, delivering to each user a unique and personalized experience while shopping online. We have the biggest database of behavioral data of the Latam market and we use it to build amazing products that deliver product recommendations, search results and targeted newsletters to millions of people everyday.

Our Data Scientists are on a mission to increment company value by applying statistics and machine learning techniques to our vast sea of data. We are looking for someone who can generate story from data and is willing to work in all stages of our pipeline.

You will help with product performance analysis and insights generation. Focusing exclusively on understanding the underlying mechanisms of how decisions can affect conversion, lift and other KPIs in our product suite.

But first, you have to prove yourself capable by overcoming this coding challenge. Since this is an opening for an intern position, we rely more on the potential shown in your solutions than on your CV to judge the candidates. So, do your best and good work!

The Challenge¶
You are given a sample of a full week of anonymized interaction events from an e-commerce.
Your task is to understand the data and show us patterns and insights to further tune our solutions for that online retailer.

For that purpose, you will have to:

Download and parse the data
Understand (research if needed) what is being asked to calculate
Write legible, reproducible and efficient code to compute those metrics
Report the results with the appropriate tables and plots

Inputs¶
Just download from this link: **HIDDEN**
20M compressed, 300M uncompressed.

All events captured are stored in JSON format. There are 4 types of events therein: page, product, search and transaction.

Below, every field of each type will be described. Many hints hidden there (Google is your friend).

Page¶
Generic pageview event.
{'eventType': 'page',            # identifies the type of event
 
 'date': '2017-06-03 18:20:53',  # timestamp of event generation
 
 'visitor': '4ce48a50-f688-11e6-aed6-b52bfa51fc22', # visitor persistent and unique id

 'deviceType': 'mobile',         # either desktop or mobile

 'osType': 'Android',            # operating system extracted from user-agent

 'osVersion': '6.0',             # operating system version

 'browserType': 'Chrome Mobile', # browser

 'browserVersion': '58.0.3029',  # browser version

 'referrerType': 'search',       # search (Google)| social (Facebook)| other| direct (no referrer)

 'utm_source': 'Partner_9',      # redacted acquisition channel partner

 'utm_medium': 'Medium_7',       # redacted acquisition medium (normally email or ads)

 'utm_campaign': 'Campaign_17',  # redacted acquisition campaign (like cart abandonment email)

 'pageType': 'subcategory',      # useful to build a conversion funnel

 'category': '319|689',          # hierarchical product category (689 is inside 319)

 'url': '7806427dfc116ddbbcb37095a477330851b0afe9',     # hashed URL

 'referrer': '18691404689e091548a691bf31b9995e8e7b9fcd' # hashed referrer
}

Product¶

Product details pageview event

{'eventType': 'product',

 'date': '2017-06-01 13:35:13',

 'visitor': '4806c650-46e8-11e7-8604-956bce5b488e',

 'deviceType': 'desktop',

 'osType': 'Windows',

 'osVersion': '7',

 'browserType': 'Chrome',

 'browserVersion': '58.0.3029',

 'referrerType': 'direct',

 'utm_source': 'Partner_4',

 'utm_medium': 'Medium_3',

 'utm_campaign': 'Campaign_3',

 'pageType': 'product',

 'category': '558|727',

 'url': '9a1cbff5cbff8666be513efeee781a71ff556f5f',

 'referrer': None,

 'product': 5639,               # product unique id

 'tags': ['216', '695', '944'], # tag codes associated with that product

 'price': 24.9,                 # product value

 'status': 'AVAILABLE'          # AVAILABLE, UNAVAILABLE or REMOVED
}

Search¶
Search results pageview event

{'eventType': 'search',

 'date': '2017-06-03 09:02:06',

 'visitor': '0eccdd90-477d-11e6-b479-450984b9b0b7',

 'deviceType': 'desktop',

 'osType': 'Windows',

 'osVersion': None,

 'browserType': 'Chrome',

 'browserVersion': '51.0.2683',

 'referrerType': None,

 'utm_source': None,

 'utm_medium': None,

 'utm_campaign': None,

 'pageType': 'search',

 'category': None,

 'url': '95a9fe3739eef4f7f38c6c0644e319ba77a1349a',

 'referrer': '5c808dff012fcbe74dc2941a7664e357db367176',

 'query': 'papel ou casa ventosa',      # obfuscated search query

 'searchItems': [6077,15749,3398,8007,  # products returned as a result of that query
                 18741,832,19311,12891, # respecting the order they were shown
                 3264,16885,20613,156,
                 11030,16204,2247,127,
                 17178,6166,268,1851,
                 2521,20810,16551,18626]
}

Transaction¶
Purchase confirmation pageview event
{'eventType': 'transaction',

 'date': '2017-06-01 17:30:54',

 'visitor': '84cc0d00-462c-11e7-93a0-39eac0558d08',

 'deviceType': 'desktop',

 'osType': 'Windows',

 'osVersion': None,

 'browserType': 'Firefox',

 'browserVersion': '53.0.0',

 'referrerType': None,

 'utm_source': None,

 'utm_medium': None,

 'utm_campaign': None,

 'pageType': 'confirmation',

 'category': None,

 'url': 'da7caa77e2729e12b32a9d7d1a324652ce2264a6',

 'referrer': '6e03ee62984224d0c0f08d4b68b819297d7f4d14',

 'order': 5545,        # unique transaction id

 'orderItems': [{      # list of products bought in that transaction
     'product': 16493, # product id
     'price': 19.9,    # product unit price
     'quantity': 1.0   # number of units bought
  },
  {'product': 1432, 'price': 20.9, 'quantity': 1.0},
  {'product': 4621, 'price': 395.01, 'quantity': 1.0}]
}


Questions¶
Using this dataset, retrieve the demanded information.

As you progress, you will notice the questions become increasingly methodologically ambigous. You will have to make decisions, and explicitely state them, on what assumptions or simplifications were made.

Always keep in mind that usefulness is the most important characteristc of any good metric.

Also, timing is more important than completeness. So, we prefer that you hand us less answers but thorough ones than all of them but half baked or belated. Go as far as you can, until the deadline.

### 1. What was the total revenue?¶
### 2. What percentage of visitors used a mobile device?¶
### 3. What search query had the highest click-through rate?¶
Among queries with at least 15 instances.
It's considered a click if the target page is a product page and its product id was in the list of search results.

### 4. When is the site most busy?¶
Do both what part of the day and what day of the week.

### 5. What is the share of revenue among categories brought by Campaign_2?¶
Use last touch attribution

### 6. Estimate the impact of unavailable products¶
Bonus points if you consider cross-elasticity
(don't fret too much. deliver something, no model is ever going to be perfect anyway)

Results delivery¶
The main deliverable is a Report with your answers.
And accompanying details, such as criteria used and neat plots.
You may use any language, tools, and cloud services you want.
But if you use Python we will like you more ;)
If you do well on the Report, we will want to see your code.
Which should be hosted on github or another online repository you can share with us.
Documentation¶
We encourage you to write down every step down the road. Design decisions, references you used, difficulties and insights, and partial results.
That way, wrapping it all up when we call you for the interview should be a breeze.

If you have any doubts, feel free to contact us.

In [137]:
import jsonlines, json
import random
import pandas
import datetime

Ok, first things first, let's answer the question:

### 1. What was the total revenue?

To answer this question, we should look for the transaction events, those events configures transactions made within the website so the revenue is directly conected to it, let's get some info about it. 

In [35]:
i = 0
with jsonlines.open('data.ndjson') as r:
    for item in r:
        if (item['eventType'] == 'transaction'):
            i+=1
    print(i)

554


In [46]:
dataTransaction = []
revenue = 0
with jsonlines.open('data.ndjson') as r:
    for item in r:
        if (item['eventType'] == 'transaction'):
            for i in item['orderItems']:
                aux = 0
                aux = i['price']
                aux *= int(i['quantity'])
                revenue += aux

In [47]:
print(revenue)

212022.26999999894


### So, the revenue is about $ 212022.26 monetary unit (probably Reais but referring as monetary unit is a safe bet)
That's answer the first question, I did the calculation thinking the data has no repeated sales or invalid ones. 
The long float number that can be seem above is probably because of how [Python solves it's equations with float](https://docs.python.org/3.4/tutorial/floatingpoint.html), and the error should be small.

If there is time I'll check this question back later, but for now let's move on to the next one.

## 2. What percentage of visitors used a mobile device?
To answer that question we need to know how to much mobile users we have, as well as what we consider a visitor. 
So, now we need to decide what we consider a visitor. This is actually a tough question since we have a loot of data about how people use the ecommerce. If we consider all the data we have, it's about 600k entries on the use of it, so it's hard to do some queries to get only single page use.

So what I call single page use? Based on the visitor field, we can select a entry of the visitor and ignore other operations he did in the ecommerce, getting a more realistic set of data of how many people use the site with mobile devices and not how much they use. The problem with this approach is that some people may use the ecommerce both with phone and with desktop (such as search something in the phone and then buy on the PC). Since those data will be obfuscated, we can try to minimize the damage by taking a random entry of the user.

One more thing, to get a more precise information we can select what kind of visitor we want to know. So I'm going to extract two informations to answer this question. First I'm going to see how much people that use the search engine of the ecommerce use it with the smartphone. To do it I'll search single page uses only on the ```eventType = 'search'```. I consider this to be a good information to get people that are already engaged with the website and, of the overall users, how much uses with smartphone. The second info is almost the same as the first, except that I'm taking the ```eventType = 'page'``` to get people that browses throw the ecommerce. Theres a lot of data on those, so I'm only taking one day, which will give me about the same number of entries of the first information. 

In [71]:
l = []
with jsonlines.open('data.ndjson') as r:
    for item in r:
        if (item['deviceType'] != None):
            if (item['eventType'] == 'search'):
                l.append(item)
    print(len(l))
    # The data is organized by date, so we need to shuffle the list to get a random value
    random.shuffle(l)
    for x,item1 in enumerate(l):
        for y,item2 in enumerate(l):
            if (item1['visitor'] == item2['visitor'] and x != y):
                l.pop(y)
                break
    ul = l            
    print(len(ul))

57737
30087


PS: The code that I'm using above to verify duplicate page views is not exactly scaleable (since is O(n²) time complexity) but for the number of entries I'm checking (less then 60k) this code runs in 5 to 10 minutes. For now, since I'm doing this code alone and in a specific data set, I'll use it that away. But if you guys have a better way to do it, maybe using ```set()``` method, I'm glad to learn.

In [72]:
count = 0
for i in ul:
    if (i['deviceType'] == 'mobile'):
        count += 1
print(count)

634


In [75]:
tam = 0
l = []
with jsonlines.open('data.ndjson') as r:
    for item in r:
        # deviceType is never null
        if (item['eventType'] == 'page'):
            tam +=1
            if (item['date'].split(' ')[0] == '2017-06-03'):
                l.append(item)
    print(tam)
    print(len(l))

527265
58426


In [77]:
random.shuffle(l)
for x,item1 in enumerate(l):
    for y,item2 in enumerate(l):
        if (item1['visitor'] == item2['visitor'] and x != y):
            l.pop(y)
            break
print(len(l))

31097


In [78]:
count = 0
for i in l:
    if (i['deviceType'] == 'mobile'):
        count += 1
print(count)

17910


In [79]:
#Percentage for first information
x = (634/30087)*100
#Percentage for second information
y = (17910/31097)*100
print(x)
print(y)

2.107222388406953
57.59398012670033


### And with those numbers, we can say that %2.10 of users use mobile to search for products and %57.59 uses to browse through the ecommerce (in one day of data analysis).

This sums up my thinking of this second question. Little more complicated than I imagine at first, but I think the information can be really useful. Let's hope on the next one.

## 3. What search query had the highest click-through rate?¶
#### Among queries with at least 15 instances. It's considered a click if the target page is a product page and its product id was in the list of search results.

Well, to begin with we better power up a little. Since now I did manage to get the answer only with json and Python json support, but from here on we should start using Pandas for better view of the data. So I'll create a data frame for each event type.

In [100]:
search = []
product = []
page = []
transaction = []
with jsonlines.open('data.ndjson') as r:
    for item in r:
        if (item['eventType'] == 'search'):
            search.append(item)
        if (item['eventType'] == 'product'):
            product.append(item)
        if (item['eventType'] == 'page'):
            page.append(item)
        if (item['eventType'] == 'transaction'):
            transaction.append(item)
dfProduct = pandas.DataFrame(product) 
dfSearch = pandas.DataFrame(search)
dfPage = pandas.DataFrame(page)
dfTransaction = pandas.DataFrame(transaction)

## 4. When is the site most busy?¶

#### Do both what part of the day and what day of the week.

In [128]:
def dayWeek(df,busyWeek):
    for index, item in df.iterrows():
        if (item['date'] != None):
            itemDate = item['date'].split(' ')[0]
            if (itemDate not in busyWeek):
                busyWeek[itemDate] = 1
            else:
                busyWeek[itemDate] += 1
    return busyWeek

In [130]:
busyWeek = {}
busyWeek = dayWeek(dfPage,busyWeek)
busyWeek = dayWeek(dfProduct,busyWeek)
busyWeek = dayWeek(dfSearch,busyWeek)
busyWeek = dayWeek(dfTransaction,busyWeek)

print(busyWeek)

{'2017-06-03': 75129, '2017-06-04': 77361, '2017-06-01': 100154, '2017-06-07': 102757, '2017-06-06': 107032, '2017-06-02': 106420, '2017-06-05': 123202}


In [144]:
def hourDay(df,dateWeek, busyDay):
    for index, item in df.iterrows():
        if (item['date'] != None):
            if (dateWeek == item['date'].split(' ')[0]):
                itemDate = item['date'].split(' ')[1]
                itemDate = datetime.datetime.strptime(itemDate, "%H:%M:%S")
                for i in range(24):
                    #first 3 values are year, month and day. Useless now.
                    if (datetime.datetime(1900,1,1,i,0,0) <= itemDate <= datetime.datetime(1900,1,1,i,59,59)):
                        if (i not in busyDay):
                            busyDay[i] = 1
                        else:
                            busyDay[i] += 1
    return busyDay

In [145]:
busyDay = {}
dataWeek = '2017-06-05'
busyDay = hourDay(dfPage,dataWeek,busyDay)
busyDay = hourDay(dfProduct,dataWeek,busyDay)
busyDay = hourDay(dfSearch,dataWeek,busyDay)
busyDay = hourDay(dfTransaction,dataWeek,busyDay)

print(busyDay)

{0: 2073, 1: 1109, 2: 816, 3: 328, 4: 347, 5: 258, 6: 692, 8: 3865, 9: 6481, 10: 7850, 11: 8343, 7: 1706, 12: 6929, 13: 7319, 14: 7776, 15: 7671, 16: 8127, 17: 7250, 18: 9664, 19: 8270, 20: 8121, 21: 7685, 22: 6149, 23: 4373}
